### PR TITLE
[ fix #835 ] Text.Regex

### LINKED ISSUE #835: Add Text.Regex
Inspiration sources:

* https://github.com/sweirich/dth/tree/master/regexp for regular expressions *with (scoped) capture*  in dependent Haskell
* https://github.com/gallais/agdarep for basic regular expressions in Agda (the optimisation described Agular and Mannaa's report does make a big difference)

It would be really nice to have regular expressions with capture (à la sed)
like Weirich does.

Bonus points: I believe we should be able to parse the input regular expression
using the shunting yard algorithm. So no need for a parser combinators library
quite yet.

### PR DESCRIPTION
**Pull request by gallais**
A late Xmas present.

### CODE REVIEW

File: src/Data/List/Relation/Ternary/Appending/Propositional.agda [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+import Data.List.Relation.Ternary.Appending.Setoid as General
+import Data.List.Relation.Ternary.Appending.Properties as Appendingₚ
+open import Data.Product using (_,_)
+open import Relation.Binary.PropositionalEquality
+  using (_≡_; setoid; refl; trans; cong; cong₂; module ≡-Reasoning)
+
+------------------------------------------------------------------------
+-- Re-export the basic combinators
+
+open General hiding (Appending; _++_; break) public
```
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Ternary/Appending/Propositional.agda:23]: Style guide says that `public` should go before `using`/`hiding`/`renaming`. Same for other files below.

File: src/Data/List/Relation/Ternary/Appending/Properties.agda [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    R : REL A B r
+    as : List A
+    bs : List B
+    cs ds : List C
+
+module _  {e} {E : Rel C e} {L′ : REL A C l} {R′ : REL B C r}
+          (LEL′ : Trans L E L′) (RER′ : Trans R E R′)
+          where
+
+  transitive : Appending L R as bs cs → Pointwise E cs ds → Appending L′ R′ as bs ds
```
Comment (MatthewDaggitt) [inline src/Data/List/Relation/Ternary/Appending/Properties.agda:34]: Hmm I think this name (and `trans` in other files) could be improved? I would expect `transitive` to have a type something like

`Appending L R as bs cs → Appending L R cs ds es→ Appending L R (as ++ bs) ds es`?

This is more like a "respects Pointwise" property. Indeed there are two other properties we could add where the pointwise equality is over the first and second list arguments right?

File: src/Text/Regex/Base.agda [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  _∣_  : (e f : Exp) → Exp
+  _∙_  : (e f : Exp) → Exp
+  _⋆   : (e : Exp) → Exp
+
+-- A regular expression has additional parameters:
+-- * should the match begin at the very start of the input?
+-- * should it span until the very end?
+
+record Regex : Set a where
+  field fromStart  : Bool
```
Comment (MatthewDaggitt) [inline src/Text/Regex/Base.agda:58]: Field names should be indented two spaces underneath the `field` keyword.

File: src/Text/Regex/Search.agda [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+... | true  | false = here
+... | false | true  = Infixₚ.fromSuffix
+... | false | false = id
+
+-- A match is a list, a proof it matches the regular expression,
+-- and a proof it is the right sort of sublist.
+record Match {s} (R : Rel (List A) s) (xs : List A) (exp : Exp)
+       : Set (a ⊔ e ⊔ r ⊔ s) where
+  constructor MkMatch
+  field list    : List A
```
Comment (MatthewDaggitt) [inline src/Text/Regex/Search.agda:69]: Indent constructors.

File: src/Text/Regex/Search.agda [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+
+open import Relation.Binary using (DecPoset)
+
+module Text.Regex.Search {a e r} (P? : DecPoset a e r) where
+
+open import Level using (_⊔_)
+open import Data.Bool.Base using (if_then_else_; true; false)
+open import Data.List.Base using (List; []; _∷_)
+open import Data.Sum.Base as Sum using (_⊎_; inj₁; inj₂; [_,_]′)
+open import Function
```
Comment (MatthewDaggitt) [inline src/Text/Regex/Search.agda:17]: `Function.Base`?

File: src/Text/Regex/Search.agda [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+  []ᴹ : ∀ {xs e} → [] ∈ e → Match (Prefix _≡_) xs e
+  []ᴹ p = MkMatch [] p []
+
+  []⁻¹ᴹ : ∀ {e} → Match (Prefix _≡_) [] e → [] ∈ e
+  []⁻¹ᴹ (MkMatch .[] p []) = p
+
+  _∷ᴹ_ : ∀ {xs e} x → Match (Prefix _≡_) xs (eat x e) → Match (Prefix _≡_) (x ∷ xs) e
+  x ∷ᴹ (MkMatch ys ys∈e\x ys≤xs) = MkMatch (x ∷ ys) (eat-sound x _ ys∈e\x) (refl ∷ ys≤xs)
+
+  _∷⁻¹ᴹ_ : ∀ {xs x e} → ¬ ([] ∈ e) →
```
Comment (MatthewDaggitt) [inline src/Text/Regex/Search.agda:93]: Very minor but why can't we use `∉` everywhere in this file?

**Review by MatthewDaggitt — COMMENTED** [review]
Looks pretty awesome :smile: I think it would probably benefit from:
- a README file?
- a `Text.Regex` module re-exporting both `Text.Regex.Base` and `Text.Regex.Search`?
- a module that specialises it to `String`s?

File: src/Data/List/Relation/Ternary/Appending.agda [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+  []       ++ rs = []++ rs
+  (l ∷ ls) ++ rs = l ∷ (ls ++ rs)
+
+  -- extract the "proper" equality split from the pointwise relation
+  break : Appending L R as bs cs → ∃₂ λ cs₁ cs₂ →
+          cs₁ List.++ cs₂ ≡ cs × Pointwise L as cs₁ × Pointwise R bs cs₂
+  break ([]++ rs) = -, -, (refl , [] , rs)
+  break (l ∷ lrs) =
+    let (_ , _ , eq , ls , rs) = break lrs in
+    -, -, (cong (_ ∷_) eq , l ∷ ls , rs)
```
[the remark this answers is not in the document]
Comment (gallais) [reply src/Data/List/Relation/Ternary/Appending.agda:57]: I'd rather not use a `with` as this means the function would get
needlessly stuck on the recursive call in proofs.

**Comment by ajrouvoet** [comment]
Random question out of interest: where does the terminology 'conicalˡ' come from?

In the ternary library I call `[]++⁻¹` the left identity of the monoidal ternary relation `Appending` and `conicalˡ` the inverse identity, for lack of a better name.

**Comment by gallais** [comment]
This was discussed in #392

**Comment by ajrouvoet** [comment]
Thank you! Awesome PR.

**Pull request merged**